Update sha256sum code to comply with OpenSSL 3.0

jeffro256 opened the pull request:
As of OpenSSL 3.0, `SHA256_Init`, `SHA256_Update`, and `SHA256_Final` are deprectaed in favor of the higher-level `EVP_*` class of functions. This causes compiler warnings, and sooner or later, will cause build errors as these functions are excluded from distro builds.

This PR does a few things:
* Updates the `sha256sum()` code to comply with OpenSSL 3.0
* Moves `sha256sum` out of `common/util` and into its own aptly named files
* Adds a bunch of error logging
* Uses the system-defined buffer size instead of a hardcoded 4096 value
* Won't throw on file error and functions are now explicitly `noexcept`
* Documentation & extra unit tests

Commit:
Update OpenSSL Code to comply with OpenSSL 3.0

As of OpenSSL 3.0, `SHA256_Init`, `SHA256_Update`, and `SHA256_Final` are deprectaed in favor of the higher-level
`EVP_*` class of functions. This causes compiler warnings, and sooner or later, will cause build errors as
these functions are excluded from distro headers.

This PR does a few things:
* Updates the `sha256sum()` code to comply with OpenSSL 3.0
* Moves `sha256sum` out of `common/util` and into its own aptly named files
* Adds a bunch of error logging
* Uses the system-defined buffer size instead of a hardcoded 4096 value
* Updates includes where necessary

jeffro256 commented:
To make this PR easier to review, you can use this as a reference example: <https://wiki.openssl.org/index.php/EVP_Message_Digests>

Commit:
Make sha256 interface uniform: i.e no throws

In the sha256sum variation which takes a filename, the function shouldn't throw
if a file fails to read, it should just return false. I checked to make sure this was
handled appropiately in the codebase. There was one instance in merge_mining.cpp that
used the other variation and didn't check for a false value, but that's unrelated.

Commit:
sha256sum: documentation and extra unit tests

jeffro256 commented on src/crypto/sha256sum.cpp:
```diff
@@ -0,0 +1,117 @@
+	bool sha256sum(const uint8_t *data, size_t len, crypto::hash &hash)
+	{
+		// Check hash size
+		if (sizeof(hash.data) < SHA256_MD_SIZE) {
```
Yes, true, it is a compile-time constant size `HASH_SIZE`. Would it be best to change it to a `static_assert`?

hyc replied:
Yeah that'd be better.

hyc commented on src/crypto/sha256sum.cpp:
```diff
@@ -0,0 +1,117 @@
+		if (!EVP_Digest(data, len, (unsigned char*) hash.data, &final_digest_size, EVP_sha256(), NULL)) {
+			MERROR("Failed to digest with OpenSSL SHA-256 EVP");
+			return false;
+		} else if (final_digest_size > sizeof(hash.data)) { // EVP_Digest overflows buffer. shouldn't ever happen, but you never know
```
These size checks also seem a bit too paranoid to me. This only goes wrong if you're not actually using a SHA256 algorithm, no? Which would imply a pretty broken OpenSSL build.

jeffro256 replied:
Yeah, they're pretty paranoid I'd say. In practice, these would only trigger if the OpenSSL implementation was broken. IIUC though, officially, up to `EVP_MAX_MD_SIZE` (36 bytes I believe)  can be written from a call to `EVP_DigestFinal_ex` which is larger than `HASH_SIZE`.

hyc replied:
If you asked for SHA256 and got 36 bytes back you didn't get what you asked for. I think we should be allowed to expect that our library dependencies are actually doing what we expect them to.

jeffro256 replied:
That's a good point, I guess we've got other things to worry about if OpenSSL is broken. I pushed a new commit with your suggestions

Commit:
sha256sum: hyc comments 1

* Use static_assert to check size of hash buffer
* Stop being so paranoid about size written from OpenSSL calls
* Above change allowed to change the functions into `noexcept` functions

hyc commented on src/crypto/sha256sum.cpp:
```diff
@@ -0,0 +1,90 @@
+		else if (!EVP_DigestInit_ex(ctx, EVP_sha256(), NULL))
+		{
+			MERROR("Failed to initialize OpenSSL SHA-256 EVP context");
+			return false;
```
This and all subsequent error returns will leak the ctx.

SChernykh replied:
No, it won't leak. `ctx` is an `std::unique_ptr` instance which will call `evp_md_ctx_freer` in destructor.

hyc commented:
I don't see the need for reorganizing sha256 to its own source file. Seems like gratuitous churn to me. Anyone else have any opinion there?

UkoeHB commented:
Somehow the ZtM2 pdf ended up in this PR.

selsta commented:
I guess the idea was to use a file with known hash, but it might be better to use something smaller so that we don't add a full megabyte to the repo.

UkoeHB commented:
FWIW, I support gratuitous changes that makes code easier to navigate. Big files that 'do everything' (e.g. `utils.cpp`) make it hard to find things and build a mental model of the code-base. It's easier to look through a list of well-organized directories and file names than it is to hunt through a giant list of forward declarations or function bodies.

Commit:
sha256sum: hyc comments 2

* sha256sum fixed leak on error
* Added licensing

Commit:
sha256sum: selsta comments

used a smaller file (~63% smaller) for unit test. It's the CLSAG IACR 2019/654 paper

UkoeHB commented on src/crypto/evp.h:
```diff
@@ -0,0 +1,83 @@
+			this->m_handle = EVP_MD_CTX_new();
+		}
+
+		/**
+		 * \brief Since recreating EVP_MD_CTXs is non-trival, copy construction is not allowed
+		 */
+		DigestContext(const DigestContext& other) = delete;
+
+		/**
+		 *\brief Since recreating EVP_MD_CTXs is non-trival, copy assignment is not allowed
+		 */
+		DigestContext& operator=(const DigestContext& other) = delete;
```
```suggestion
		/**
		 * \brief Since recreating EVP_MD_CTXs is non-trival, copies and moves are not allowed.
		 */
		DigestContext& operator=(DigestContext&&) = delete;
```

The so-called 'desdemova' trick to disable both copies and moves.

hyc commented:
Adding a new class to wrap a structure that is never used outside of the functions that create it seems pretty bloaty to me.

jeffro256 commented:
@hyc Yeah I agree. Currently working on finding the least ugly way to express this.

jeffro256 commented:
Okay I made it a `unique_ptr` with a deleter. Its really small and since this is the only point where EVP *would* be used in the main codebase, AFAIC, I think this is a good solution.

selsta commented:
This alternative patch was suggested on IRC: https://paste.debian.net/hidden/74158896/

jeffro256 commented:
@selsta My PR differs from this diff in a couple of ways:

1. Logs what causes the errors at each step to help debug
2. Doesn't throw on file error, and can be used in `noexcept` contexts
3. Interface is documented
4. Logic is moved to aptly named file
5. Changes how the file is buffered

Commit:
sha256sum: selsta comment 2

Change the unique_ptr deleter to a function pointer to EVP_MD_CTX_free instead of a lambda since it looks better

selsta commented:
Sure, I meant it's an alternative patch to do "Update sha256sum code to comply with OpenSSL 3.0 ". Your code obviously does more.

Commit:
sha256sum: selsta comment 3

tabs -> spaces

jeffro256 commented:
This PR seems to have consumed a lot of dev time, so I want to hold a vote to see if I should close it. Thumbs up to keep open, thumbs down to close.

selsta commented:
I mean we will need a patch to `comply with OpenSSL 3.0`, whether it's this PR or a different one.

The easiest way we can get it merged is by keeping the diff as minimal as possible, adding tests and documentation is also good in my opinion, it doesn't make things more difficult to review.

All the other changes are personal preference which will result in discussion and we had an unwritten rule about not rewriting code for personal preference.

Commit:
sha256sum: forward sha266sum from util.h

Reduces git diff and other files don't have to change includes

Commit:
sha256sum: add unit test for file failure

jeffro256 commented:
@hyc Now the only existing modified files in `src/` are `util.cpp`, `util.h`, and `crypto/CMakeLists.txt` and those diffs are really easy to review. `sha256sum.cpp` and `sha256sum.h` remain the same.

selsta commented:
This would be a mix between both patches, I added you as the author: https://github.com/selsta/monero/commit/32ec76faab82baccede4bac85ead0f9d9ae1d99c

jeffro256 commented:
Continue at #8340
